compiler/noirc_frontend/src/hir/resolution/import.rs

```diff
@@ -218,7 +257,72 @@ fn resolve_external_dep(
         is_prelude: false,
     };
 
-    let dep_def_map = def_maps.get(&dep_module.krate).unwrap();
+    resolve_path_to_ns(&dep_directive, dep_module.krate, importing_crate, def_maps, allow_contracts)
+}
 
-    resolve_path_to_ns(&dep_directive, dep_def_map, def_maps, allow_contracts)
+// Issue an error if the given private function is being called from a non-child module, or
+// if the given pub(crate) function is being called from another crate
+fn check_can_reference_function(
+    def_maps: &BTreeMap<CrateId, CrateDefMap>,
+    importing_crate: CrateId,
+    current_module: ModuleId,
+    target_module: ModuleId,
+    visibility: ModuleVisibility,
+    ident: &Ident,
+) -> Result<(), PathResolutionError> {
+    let same_crate = target_module.krate == importing_crate;
+    let current_module = current_module.local_id;
+
+    match visibility {
+        ModuleVisibility::Public => Ok(()),
+        ModuleVisibility::Private
+            if !same_crate
+                || !module_parent_of_target(
+                    def_maps,
+                    importing_crate,
+                    target_module.local_id,
+                    current_module,
+                ) && !module_descendent_of_target(
+                    def_maps,
+                    importing_crate,
+                    target_module.local_id,
+                    current_module,
+                ) =>
+        {
+            Err(PathResolutionError::Private(ident.clone()))
+        }
+
+        ModuleVisibility::PublicCrate if !same_crate => {
+            Err(PathResolutionError::Private(ident.clone()))
+        }
+
+        _ => Ok(()),
+    }
+}
+
+// Returns true if `current` is a (potentially nested) child module of `target`.
+// This is also true if `current == target`.
+fn module_descendent_of_target(
+    def_maps: &BTreeMap<CrateId, CrateDefMap>,
+    krate: CrateId,
+    target: LocalModuleId,
+    current: LocalModuleId,
+) -> bool {
+    if current == target {
+        return true;
+    }
+
+    def_maps[&krate].modules[current.0]
+        .parent
+        .map_or(false, |parent| module_descendent_of_target(def_maps, krate, target, parent))
+}
+
+// Returns true if `target` is a direct child module of `current`.
+fn module_parent_of_target(
```

Comment:
This is working around private modules not making their contents available to their parent module (i.e. importing public functions from a private module will fail).

Seems like we should add `pub(super)` functionality and make this the default for `mod foo;`